### PR TITLE
Use `ParameterReferences` type for `ParameterTable` values

### LINKED ISSUE #6725: `QuantumCircuit._check_dup_param_spec` becomes slow for many gates reusing few parameters
### What is the expected enhancement?

When building a parameterized circuit where a large number of gates re-use a comparatively small number of parameters, `QuantumCircuit._check_dup_param_spec` becomes a performance bottleneck.

For example, `circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit` takes 90s to build a 131k gate circuit with 8 distinct parameters, vs. <5s to build a 131k gate circuit with anything more than 128 parameters:

https://qiskit.github.io/qiskit/#circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit?x-axis=number%20of%20params

Since the structure of the `ParameterTable` is dict mapping `Parameter` instances to lists of `Instruction`s, each call to `_check_dup_param_spec` must walk the entire `Instruction` list to look for a match. One option would be to convert the list to a dict, which could be efficiently checked for membership.

### PR DESCRIPTION
**Pull request by kevinhartman**
### Summary
Changes `ParameterTable` value type from `List[(Instruction, int)]` to a new collection type, `ParameterReferences`, which is set-like, and **no longer seq-like**.

`ParameterReferences`:

> A set of instruction parameter slot references. Items are expected in the form `(instruction, param_index)`. Membership testing is overriden such that items that are otherwise value-wise equal are still considered distinct if their `instruction`s are referentially distinct.

### Details and comments
It's not actually necessary to update existing uses of `ParameterTable` or its values, but usage is updated in `QuantumCircuit` for performance (mostly to avoid creating lists that are then immediately spooled into `ParameterReferences`). See `ParameterReferences` docstring for more information.

Fixes #6725. `ParameterReferences` is set-like, so there's no need to perform a linear search in `_update_parameter_table` to check existing references for every instruction parameter.

### CODE REVIEW

**Comment by kevinhartman** [comment]
Tagging @jakelishman esp. for Python best practice insights :)

File: qiskit/circuit/parametertable.py [hunk 1]
```diff
@@ -12,43 +12,140 @@
...
+            k = self._instance_key(ref)
+            if k not in self._instance_ids:
+                self._data.append(ref)
+                self._instance_ids.add(k)
+
+    @classmethod
+    def from_iterable(cls, iterable):
+        """Create a new instance from an iterable of ``(instruction, param_index)``
+        tuples."""
+        return cls(*iterable)
```
Comment (jakelishman) [inline qiskit/circuit/parametertable.py:52]: This is entirely minor, but I'm not really sure why `__init__` takes a variadic argument and this "from iterable" form isn't just the only (and default) constructor.

File: qiskit/circuit/parametertable.py [hunk 2]
```diff
@@ -12,43 +12,140 @@
...
 """
-import warnings
 import functools
-from collections.abc import MutableMapping, MappingView
+import warnings
+from collections.abc import MappingView, MutableMapping, Set
+from typing import MutableSequence
+
+
+class ParameterReferences(MutableSequence, Set):
```
Comment (jakelishman) [inline qiskit/circuit/parametertable.py:21]: Is there a reason to maintain the insertion order here other than just backwards compatibility?  The mutable set interface with a custom key seems to be much more natural for what we need.

File: qiskit/circuit/parametertable.py [hunk 3]
```diff
@@ -12,43 +12,140 @@
...
+
+    def __contains__(self, x) -> bool:
+        return self._instance_key(x) in self._instance_ids
+
+    def __repr__(self) -> str:
+        return repr(self._data)
 
-from .instruction import Instruction
+    def insert(self, index, value) -> None:
+        raise NotImplementedError("Position is dictated by insertion order.")
```
Comment (jakelishman) [inline qiskit/circuit/parametertable.py:78]: I'm not sure why you'd forbid this - it just causes you to violate the `MutableSequence` interface, and I'm not sure why it'd hurt to allow an override.  That said, I'm not sure if we need to maintain the `MutableSequence` part of this anyway.

File: test/python/circuit/test_parameters.py [hunk 4]
```diff
@@ -1782,6 +1782,184 @@ def test_parameter_symbol_equal_after_ufunc(self):
         self.assertEqual(phi._parameter_symbols, cos_phi._parameter_symbols)
 
 
+class TestParameterReferences(QiskitTestCase):
+    """Test the ParameterReferences class."""
+
+    def test_equal_inst_diff_instance(self):
+        """Different value equal instructions are treated as distinct."""
+        from qiskit.circuit.library.standard_gates.rz import RZGate
```
Comment (jakelishman) [inline test/python/circuit/test_parameters.py:1790]: There's usually no need to hide imports in tests, and that'll cut down the import repetition - if anything is shadowing the name `RZGate` in the test, we're already screwed haha.

**Review by jakelishman — COMMENTED** [review]
This looks pretty sensible, and looks like it should reduce a potential quadratic complexity in circuit depth when using parameters.  Do you have benchmarks for improvement for it to hand?

File: qiskit/circuit/parametertable.py (hunk 2 above)
Comment (kevinhartman) [reply qiskit/circuit/parametertable.py:21]: Currently, there are a number of places in Terra that depend on parameter references being indexable, so I assumed it was something we should preserve. I can take a look at those callsites again and see if I can rewrite the logic if you think that'd be better. Thoughts? :)

File: qiskit/circuit/parametertable.py [hunk 5]
```diff
@@ -12,43 +12,140 @@
...
+    def from_iterable(cls, iterable):
+        """Create a new instance from an iterable of ``(instruction, param_index)``
+        tuples."""
+        return cls(*iterable)
+
+    def __getitem__(self, index):
+        return self._data[index]
+
+    def __setitem__(self, index, ref):
+        raise NotImplementedError("Position is dictated by insertion order.")
```
[the remark this answers is not in the document]
Comment (kevinhartman) [reply qiskit/circuit/parametertable.py:58]: It gets kind of tricky. The problem is that we need to preserve the set-like property if we were to allow an update operation like this.

Consider the following example, where `PR` is both set and seq-like.
```py
s = PR("a", "b", "c")

s[1] = "a"
print(s[1]) # uhh?
```

What should happen? It seems clear that we want to replace `"b"` with `"a"`, but `"a"` is already at position 0. So we need to remove it. If we do, the new `"a"` will actually live at position 0, and `s[1]` will actually be `"c"`, which seems confusing / just wrong. We could instead keep a placeholder `None` where `"a"` used to be, but now we'll have `None`s in our sequence.

Tl;dr, I couldn't think of a reasonable way to support index-based set while keeping `ParameterReferences` set-like.

File: qiskit/circuit/parametertable.py (hunk 1 above)
Comment (kevinhartman) [reply qiskit/circuit/parametertable.py:52]: IIRC, I did this to support syntax like `ParameterReferences(ref1, ref2)`, which is used in a couple of places. I believe there may be another subtlety that I'm forgetting in the choice of adding a named constructor like this, having to do with not being able to write a single `__init__` in a way that can disambiguate iterables from references.

I remember taking inspiration from [itertools.chain](https://docs.python.org/3/library/itertools.html#itertools.chain).

Let me know if you'd like me to change this.

File: qiskit/circuit/parametertable.py (hunk 3 above)
Comment (kevinhartman) [reply qiskit/circuit/parametertable.py:78]: This is forbidden for the same reason as is set by index. I agree that it's less than ideal to violate `MutableSequence`. Perhaps we should get rid of index-based access from existing places in Terra and drop the seq-like property here.

File: test/python/circuit/test_parameters.py (hunk 4 above)
Comment (kevinhartman) [reply test/python/circuit/test_parameters.py:1790]: Heh. Good point. I just copied this from some other nearby tests, assuming they had a good reason for doing it.

**Comment by kevinhartman** [comment]
> Do you have benchmarks for improvement for it to hand?

I don't! I should have some time to get some numbers in the coming week, and will report back.

File: qiskit/circuit/parametertable.py (hunk 2 above)
Comment (jakelishman) [reply qiskit/circuit/parametertable.py:21]: Is there anywhere that _logically_ need to access by index?  Off the top of my head, I wouldn't have expected so - I can't think of why you'd need an ordering to them, and there's a good chance that the iterable form is causing performance issues in things like duplication checking (I know some of that happens in `QuantumCircuit` for sure, which is presumably the point of this PR).  If there's nothing that needs the sequence, it'd be good to drop it entirely, I think.  In that case, it would become `MutableSet`, not just `Set`.

File: qiskit/circuit/parametertable.py (hunk 1 above)
Comment (jakelishman) [reply qiskit/circuit/parametertable.py:52]: I think `itertools.chain` is perhaps solving a slightly different problem - in that, it's common to want to construct the object with "hardcoded" arguments, like
```python
for key in itertools.chain(left_keys, right_keys):
    ...
```
where it's convenient syntactically to not require the extra `[]` or `()` in the constructor, and it's generic over its inputs, so it's valid to chain together iterables of iterables (i.e. they can't infer from the input, but you could if you wanted to).  `chain` also needs to support infinite/lazy iterable inputs, which is where `from_iterable` really becomes necessary.

In this, the only situation I can think of where you'd create one with a "hardcoded" input is the singleton case, but for all the rest you want to iterate through an unknown number of references to build it up, so I'd perhaps optimise its syntax around that.

File: qiskit/circuit/parametertable.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jakelishman) [reply qiskit/circuit/parametertable.py:58]: Yeah, that's a very valid point.  I'd mostly be trying to get rid of the need for indexed access - I think/hope that should be possible.

File: qiskit/circuit/parametertable.py (hunk 2 above)
Comment (kevinhartman) [reply qiskit/circuit/parametertable.py:21]: Agreed—it'd be great to just have this be a `MutableSet`. I'll look into it and report back. Originally, I took the shortcut of assuming there was a logical reason to index into the sequence since to me, it seemed like an odd thing to do otherwise :).

**Comment by kevinhartman** [comment]
@jakelishman, I found at least one place that depends on `ParameterReferences` being seq-like. Here, for each reference, we need to look up the corresponding copy of the same reference in a new circuit(s).

https://github.com/Qiskit/qiskit-terra/blob/5ff5757a762a677baa7db2fde09559ee87a0c8dc/qiskit/opflow/gradients/circuit_gradients/param_shift.py#L208-L218

I'm not really sure if there's a good way to do this if `ParameterReferences` is no longer ordered, since `ParameterReferences` might (by design) contain multiple references that are equivalent, but not the same. We can't do any sort of lookup since we're dealing with a deep copy in the new circuit.

Any ideas on how we should proceed?

**Comment by kevinhartman** [comment]
Here are the `asv` performance results for `time_build_parameterized_circuit`, which show a significant speedup for large numbers of gates that re-use a comparatively small set of parameters.

```
  84.6±1ms         74.6±1ms     0.88  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 2048, 8)
1.47±0.03s       1.20±0.01s     0.82  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 32768, 128)
   484±6ms          307±3ms     0.63  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8192, 8)
14.8±0.06s       4.91±0.03s     0.33  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 131072, 128)
4.23±0.04s       1.23±0.01s     0.29  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 32768, 8)
   2.15±0m       5.02±0.03s     0.04  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 131072, 8)
```

File: qiskit/circuit/quantumcircuit.py [hunk 6]
```diff
@@ -1289,31 +1289,19 @@ def _update_parameter_table(self, instruction: Instruction) -> Instruction:
...
+                    self._parameter_table[parameter] = ParameterReferences(
+                        (instruction, param_index)
+                    )
 
                     # clear cache if new parameter is added
                     self._parameters = None
 
         return instruction
 
-    def _check_dup_param_spec(
```
Comment (kevinhartman) [inline qiskit/circuit/quantumcircuit.py:1306]: I should probably write a release note for the expected performance improvement.

**Comment by ewinston** [comment]
Is there an asv test for a large number of gates with a large number of parameters?

**Comment by jakelishman** [comment]
@ewinston: Yeah, the same benchmark suite Kevin showed goes up to 131,072 gates x 131,072 parameters, but I'm guessing he just didn't show them here because there wasn't much change (and in total there's ~36 benchmarks in that suite).

**Comment by jakelishman** [comment]
@kevinhartman: sorry, I haven't had chance to look properly, but I feel like that code block there is logically just trying to match up the instructions between circuits, including from a main circuit.  I think it might be possible to modify the algorithm in that location a little such that it iterates over a index into the `QuantumCircuit.data` list, and then uses that index for each circuit to retrieve the instruction tuple, from which it can access the parameter table.  I'm not 100% sure, but I'd be hopeful we can change things on the algorithm side rather than requiring `ParameterReferences` to be sequence-like, because in the end, we don't really want circuit data to have a total ordering - the instruction flow is logically a graph, not a sequence.

(I get that my proposed solution uses the total ordering of `data`, but I'm mostly just pushing the issue down the line, because when we're ready to move the backing data structure of `QuantumCircuit.data` to a graph-like one, we'll have to solve these problems far more generally across Terra.)

**Comment by kevinhartman** [comment]
> I think it might be possible to modify the algorithm in that location a little such that it iterates over a index into the QuantumCircuit.data list, and then uses that index for each circuit to retrieve the instruction tuple, from which it can access the parameter table.

Hmm, alright! That's helpful. I'll give it a try.

> (I get that my proposed solution uses the total ordering of data, but I'm mostly just pushing the issue down the line, because when we're ready to move the backing data structure of QuantumCircuit.data to a graph-like one, we'll have to solve these problems far more generally across Terra.)

I agree completely that it'd be nice to eliminate maintaining an arbitrary ordering of parameter references, at the very least.

File: qiskit/circuit/parametertable.py [hunk 7]
```diff
@@ -12,43 +12,113 @@
...
+
+    def __iter__(self):
+        for (_, idx), instruction in self._instance_ids.items():
+            yield (instruction, idx)
+
+    def __contains__(self, x) -> bool:
+        return self._instance_key(x) in self._instance_ids
+
+    def __repr__(self) -> str:
+        return f"ParameterReferences({repr(list(iter(self)))})"
```
Comment (jakelishman) [inline qiskit/circuit/parametertable.py:51]: Incredibly minor, but `list(self)` should already do the right thing - no need to manually call `iter`.

File: qiskit/circuit/quantumcircuit.py (hunk 6 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py:1288]: I think you still might need to do this?  The benchmarks you showed earlier in the review process look :rocket: already.

File: test/python/circuit/test_parameters.py [hunk 8]
```diff
@@ -1783,6 +1781,137 @@ def test_parameter_symbol_equal_after_ufunc(self):
...
+
+        # check copy not modified
+        self.assertNotIn(ref2, refs_copy)
+        self.assertEqual(list(refs_copy), [ref0, ref1])
+
+        refs_copy.add(ref3)
+
+        # check original not modified
+        self.assertNotIn(ref3, refs)
+        self.assertEqual(list(refs), [ref0, ref1, ref2])
```
Comment (jakelishman) [inline test/python/circuit/test_parameters.py:1855]: Iteration order?

Comment (jakelishman) [inline test/python/circuit/test_parameters.py:1855]: This test says it's testing for shallow copies, but then it uses logical equality checks rather than referential equality tests when comparing the values stored.  If we fully want to ensure that it's not just a copy (which is currently tested), but specifically a _shallow_ copy, then we probably need the referential equality tests.

File: test/python/circuit/test_parameters.py [hunk 9]
```diff
@@ -1783,6 +1781,137 @@ def test_parameter_symbol_equal_after_ufunc(self):
...
+
+        mapping = {p1: ParameterReferences((ref0, ref1)), p2: ParameterReferences((ref2,))}
+
+        table = ParameterTable(mapping)
+
+        # make sure editing mapping doesn't change `table`
+        del mapping[p1]
+
+        self.assertEqual(list(table[p1]), [ref0, ref1])
+        self.assertEqual(list(table[p2]), [ref2])
```
Comment (jakelishman) [inline test/python/circuit/test_parameters.py:1879]: Iteration order?

File: test/python/circuit/test_parameters.py [hunk 10]
```diff
@@ -1783,6 +1781,137 @@ def test_parameter_symbol_equal_after_ufunc(self):
...
+
+        ref0 = (RZGate(p1), 0)
+        ref1 = (RZGate(p1), 0)
+        ref2 = (RZGate(p1), 0)
+
+        table = ParameterTable(
+            {
+                p1: ParameterReferences((ref0, ref1)),
+            }
+        )
```
Comment (jakelishman) [inline test/python/circuit/test_parameters.py:1909]: Nitpicky, but I think this should fit on one line, it's just the trailing comma on line 1907 that puts black in "force individual lines" mode for the dict comprehension.

File: test/python/circuit/test_parameters.py [hunk 11]
```diff
@@ -1783,6 +1781,137 @@ def test_parameter_symbol_equal_after_ufunc(self):
...
+
+        table[p1] = (ref2, ref1, ref0)
+        self.assertEqual(list(table[p1]), [ref2, ref1, ref0])
```
Comment (jakelishman) [inline test/python/circuit/test_parameters.py:1912]: Insertion order?

File: test/python/circuit/test_parameters.py [hunk 12]
```diff
@@ -1783,6 +1781,137 @@ def test_parameter_symbol_equal_after_ufunc(self):
...
+        theta = Parameter("theta")
+        ref0 = (RZGate(theta), 0)
+        ref1 = (RZGate(theta), 0)
+        ref2 = (RZGate(theta), 0)
+
+        refs = ParameterReferences((ref0,))
+
+        refs |= ParameterReferences((ref0, ref1, ref2, ref1, ref0))
+
+        self.assertEqual(list(refs), [ref0, ref1, ref2])
```
[the remark this answers is not in the document]
Comment (kevinhartman) [reply test/python/circuit/test_parameters.py:1832]: Ah, I'm glad you noticed this! It's a left-over from when `ParameterReferences` was also seq-like in my first draft.

File: qiskit/opflow/gradients/circuit_gradients/param_shift.py [hunk 13]
```diff
@@ -205,17 +205,22 @@ def _parameter_shift(
...
                 mshift_op = deepcopy(operator)
 
                 # We need the circuit objects of the newly instantiated operators
                 pshift_circ = self.get_unique_circuits(pshift_op)[0]
                 mshift_circ = self.get_unique_circuits(mshift_op)[0]
 
-                pshift_gate = pshift_circ._parameter_table[param][m][0]
-                mshift_gate = mshift_circ._parameter_table[param][m][0]
+                pshift_gate = pshift_circ.data[m][0]
+                mshift_gate = mshift_circ.data[m][0]
```
[the remark this answers is not in the document]
Comment (kevinhartman) [reply qiskit/opflow/gradients/circuit_gradients/param_shift.py:223]: Yeah... I don't have a complete understanding of what this code is doing. It doesn't seem like it takes into account that an instruction might reference multiple parameters (and hence multiple parameter reference entries might contain the same instruction). I tried to stay as close to the original logic to avoid regressions.

File: qiskit/circuit/parametertable.py [hunk 14]
```diff
@@ -12,43 +12,113 @@
...
+        return (id(ref[0]), ref[1])
+
+    def __init__(self, refs):
+        self._instance_ids = {}
+
+        for ref in refs:
+            if not isinstance(ref, tuple) or len(ref) != 2:
+                raise ValueError("refs must be in form (instruction, param_index)")
+            k = self._instance_key(ref)
+            self._instance_ids[k] = ref[0]
```
[the remark this answers is not in the document]
Comment (kevinhartman) [reply qiskit/circuit/parametertable.py:38]: Done in [acc3a89](https://github.com/Qiskit/qiskit-terra/pull/7408/commits/acc3a895869ee661b6326dbdf1769550eb57697b).

File: test/python/circuit/test_parameters.py (hunk 10 above)
Comment (kevinhartman) [reply test/python/circuit/test_parameters.py]: Fixed in 1ff03a5.

File: test/python/circuit/test_parameters.py (hunk 9 above)
Comment (kevinhartman) [reply test/python/circuit/test_parameters.py]: Fixed in 1ff03a5.

File: test/python/circuit/test_parameters.py [hunk 15]
```diff
@@ -1783,6 +1781,137 @@ def test_parameter_symbol_equal_after_ufunc(self):
...
+
+        refs = ParameterReferences(((gate1, 0), (gate2, 0)))
+
+        # test __contains__
+        self.assertIn((gate1, 0), refs)
+        self.assertIn((gate2, 0), refs)
+
+        gate_ids = {id(gate1), id(gate2)}
+        self.assertTrue(all(id(gate) in gate_ids for gate, _ in refs))
+        self.assertTrue(all(idx == 0 for _, idx in refs))
```
[the remark this answers is not in the document]
Comment (kevinhartman) [reply test/python/circuit/test_parameters.py]: Fixed in 1ff03a5.

File: qiskit/circuit/parametertable.py [hunk 16]
```diff
@@ -12,43 +12,113 @@
...
 
-    def __init__(self, *args, **kwargs):
-        """
-        the structure of _table is,
-           {var_object: [(instruction_object, parameter_index), ...]}
+    def __init__(self, mapping=None):
+        """Create a new instance, initialized with ``mapping`` if provided.
+
+        Args:
+            mapping (Mapping[Parameter, ReferenceListing]):
```
[the remark this answers is not in the document]
Comment (kevinhartman) [reply qiskit/circuit/parametertable.py]: Fixed in 1ff03a5.

File: qiskit/circuit/parametertable.py [hunk 17]
```diff
@@ -12,43 +12,113 @@
...
+        k = self._instance_key(value)
+        self._instance_ids[k] = value[0]
+
+    def discard(self, value):
+        k = self._instance_key(value)
+        self._instance_ids.pop(k, None)
+
+    def copy(self):
+        """Create a shallow copy."""
+        return ParameterReferences(iter(self))
```
[the remark this answers is not in the document]
Comment (kevinhartman) [reply qiskit/circuit/parametertable.py]: Fixed in 1ff03a5.

File: qiskit/circuit/parametertable.py (hunk 7 above)
Comment (kevinhartman) [reply qiskit/circuit/parametertable.py]: Fixed in 1ff03a5.

File: qiskit/circuit/quantumcircuit.py (hunk 6 above)
Comment (kevinhartman) [reply qiskit/circuit/quantumcircuit.py:1288]: Done in b5a28d5.

**Review by jakelishman — APPROVED** [review]
Looks good to me, and the speed-ups are great!  Thanks!
